docs/Contributing/Blueprints/Troubleshooting-Hosts.md

```diff
@@ -0,0 +1,284 @@
+processes the write request).
+- The algorithm sends a "trouble query" as a "request stamp" in the `distributed/read`.
+This way Fleet can tie the queries sent in the `distributed/read` to the results received in the
+`distributed/write`. (Such stamp contains the hash of the queries.)
+- State storage for each host with issues:
+	- To not generate more load we can define a new column in `hosts` called something like
+	`trouble_queries`. Given that we load the host by selecting `hosts` on every request, this
+	should cause no extra performance penalty on every request).
+	- Alternatively, we could use Redis to store such state, but it would add a Redis read request
+	on every `distributed/read` request (NOTE: we are already doing a Redis request on the
+	`distributed/read`, for retrieving live queries).
+- We need to use and store the hashes of the queries because queries can change and new queries can be
+added in between read and write requests (e.g. policy queries are editable by users and new ones
+can be assigned to hosts).
+- The algorithm performs a sort of binary search of the problematic query.
+- The added performance penalty of the algorithm is in `distributed/read`, there's now an extra
+write to `hosts` table (only on hosts with issues).
+
+Following is the rough/untested pseudo-code for the `distributed/read` and `distributed/write` to
+implement the binary search. The algorithm keeps a JSON marshalled `TroubleQueriesState` for each problematic host in the `hosts` table.
+	
+#### `distributed/read`
+
+```go
+type TroubleQueriesState struct {
+	BatchIndex int
+	BatchSize  int
+	// Queries holds the suspicious queries (currently in search for culprits).
+	Queries    []QueryWithHash
+	// Denylisted holds the queries that have been confirmed to cause issues in the host.
+	Denylisted []QueryWithHash
+}
+
+type QueryWithHash struct {
+	Query string
+	Hash  string
+}
+
+func determineQueriesToSend(host *fleet.Host, queriesToSend map[string]string) map[string]string {
+	hostNotResponding := svc.detectHostNotResponding(host)
+
+	// TODO(lucas): Check host.TroubleQueries.Denylisted, and not include these in the response.
+
+	switch {
+
+	case !hostNotResponding && host.TroubleQueries == "":
+
+		//
+		// Nothing to do, host is responding and there are no "trouble queries".
+		//
+
+		return queriesToSend
+
+	case hostNotResponding && host.TroubleQueries == "":
+
+		//
+		// This is the first time Fleet detects the host has issues.
+		//
+		
+		// We assume hostDetailQueries never have performance issues.
+		nonDetailQueries := excludeQueries(queriesToSend, hostDetailQueries)
+		
+		queriesWithHash := withHash(nonDetailQueries)
+		host.TroubleQueries = TroubleQueriesState{
+			BatchIndex: 0,
+			BatchSize: len(queriesToSend)/2,
+			Queries: queriesWithHash,
+		}
+		svc.ds.UpdateHost(ctx, host)
+
+		queriesToSend := excludeQueries(queriesToSend, host.TroubleQueries.Queries)
```

Comment:
This should be moved above the `host.TroubleQueries = ...` so that the batch size is correct. We might want to also consider instead storing a `BatchDivider` and we can start 2, 3, 4, etc.